### PR TITLE
Problem: logs in callback contract are lost

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #1231

Solution:
- re-emit those logs

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

## Summary by CodeRabbit

Bug Fixes:
- Fixed issues related to the default chain ID behavior and redundant parsing of chain ID from genesis, improving system stability.
- Addressed a problem with the memory store in the versiondb multistore, enhancing data storage efficiency.

New Features:
- Enhanced the `onPacketResult` function to handle the response from the EVM call and add logs to the stateDB, improving data tracking.
- Added a feature to re-emit logs in a callback contract, improving log management and traceability.

Tests:
- Added a `print` statement in the integration tests to display the logs of the `OnPacketResult` events, aiding in debugging and test validation.
- Introduced a new test function `test_sc_call` and a helper function `assert_logs_equal` for more comprehensive testing.

Refactor:
- Modified the context in the `exec` function to carry additional information related to the state DB, enhancing context data management.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 - [#1217](https://github.com/crypto-org-chain/cronos/pull/1217) Use the default chain-id behavour in sdk.
 - [#1216](https://github.com/crypto-org-chain/cronos/pull/1216) Update ethermint to fix of avoid redundant parse chainID from gensis when start server.
 - [#1230](https://github.com/crypto-org-chain/cronos/pull/1230) Fix mem store in versiondb multistore.
+- [#]() Re-emit logs in callback contract.
```
Comment (yihuang) [inline CHANGELOG.md:11]:
```suggestion
- [#1233](https://github.com/crypto-org-chain/cronos/pull/1233) Re-emit logs in callback contract.
```

File: integration_tests/test_ica_precompile.py [hunk 2]
```diff
@@ -302,3 +302,5 @@ def submit_msgs_ro(func, str):
     assert expected_seq == last_seq
     assert status == Status.FAIL
     assert cli_host.balance(ica_address, denom=denom) == balance
+
+    print(tcontract.events.OnPacketResult().get_logs())
```
Comment (yihuang) [inline integration_tests/test_ica_precompile.py:306]: TODO, assert the logs

File: x/cronos/keeper/keeper.go [hunk 3]
```diff
@@ -298,8 +299,17 @@ func (k Keeper) onPacketResult(
...
-	_, _, err = k.CallEVM(ctx, &contractAddr, data, big.NewInt(0), gasLimit)
-	return err
+	_, rsp, err := k.CallEVM(ctx, &contractAddr, data, big.NewInt(0), gasLimit)
+	if err != nil {
+		return err
+	}
+
+	if stateDB, ok := ctx.Value("statedb").(vm.StateDB); ok {
+		for _, l := range rsp.Logs {
+			stateDB.AddLog(l.ToEthereum())
```
Comment (thomas-nguy) [inline x/cronos/keeper/keeper.go:309]: Interesting, so that mean we will be ablet to get the CRC20 logs for the indexer for ibc transaction?

Comment (yihuang) [reply x/cronos/keeper/keeper.go:309]: yea

File: integration_tests/test_ica_precompile.py (hunk 2 above)
Comment (mmsqe) [reply integration_tests/test_ica_precompile.py:306]: yea, we might need get_event_data from the logs

File: x/cronos/keeper/precompiles/utils.go [hunk 4]
```diff
@@ -45,6 +46,7 @@ func exec[Req any, PReq interface {
 	var res Resp
 	if err := stateDB.ExecuteNativeAction(contract, converter, func(ctx sdk.Context) error {
 		var err error
+		ctx = ctx.WithValue(types.StateDBContextKey, stateDB)
```
Comment (mmsqe) [inline x/cronos/keeper/precompiles/utils.go:49]: we might add in [ica](https://github.com/crypto-org-chain/cronos/blob/8a3e3d4c2dc64e4942c505e4ded23c4263a6423b/x/cronos/keeper/precompiles/ica.go#L183) as well, but I think callback [ctx](https://github.com/cosmos/ibc-go/blob/17cf1260a9cdc5292512acc9bcf6336ef0d917f4/modules/apps/callbacks/ibc_middleware.go#L246) is different after go through http call

Comment (yihuang) [reply x/cronos/keeper/precompiles/utils.go:49]: just need to use `ibc_rly` setup.

**Comment by yihuang** [comment]
integration test passed.

File: x/cronos/keeper/precompiles/utils.go (hunk 4 above)
Comment (mmsqe) [reply x/cronos/keeper/precompiles/utils.go:49]: but it's change to call [relayer](https://github.com/crypto-org-chain/cronos/blob/24750f14e34eb3e368e2dfe8189c54b7865238bc/integration_tests/configs/ibc_rly.jsonnet#L6) precompile which use [exec](https://github.com/crypto-org-chain/cronos/blob/8a3e3d4c2dc64e4942c505e4ded23c4263a6423b/x/cronos/keeper/precompiles/utils.go#L21) from utils right? do we need call ctx.WithValue in ica?

**Review by JayT106 — APPROVED** [review]
LGTM

**Review by mmsqe — APPROVED** [review]

**Pull request merged**